crates/holochain/src/core/sys_validate.rs

```diff
@@ -391,3 +398,317 @@ pub fn check_update_reference(
+/// to our incoming_dht_ops_workflow.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_store_entry<F>(
+    hash: &HeaderHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_entry(source).await?;
+    Ok(())
+}
+
+/// If we are not holding this entry then
+/// retrieve any element at this EntryHash
+/// and send it as a StoreEntry DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Note this is different to check_and_hold_store_entry
+/// because it gets the Element via an EntryHash which
+/// means it will be any Element.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_any_store_entry<F>(
+    hash: &EntryHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_entry(source).await?;
+    Ok(())
+}
+
+/// If we are not holding this header then
+/// retrieve it and send it as a StoreElement DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_store_element<F>(
+    hash: &HeaderHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_element(source).await?;
+    Ok(())
+}
+
+/// Allows you to send an op to the
+/// incoming_dht_ops_workflow if you
+/// found it on the network and were supposed
+/// to be holding it.
+#[derive(derive_more::Constructor)]
+pub struct IncomingDhtOpSender {
+    env: EnvironmentWrite,
+    sys_validation_trigger: TriggerSender,
+    check_level: CheckLevel,
+}
+
+impl IncomingDhtOpSender {
+    /// Sends the op to the workflow if it was found on the network
+    async fn send_op(
+        self,
+        source: Source,
+        make_op: fn(Element) -> Option<(DhtOpHash, DhtOp)>,
+    ) -> SysValidationResult<()> {
+        if let Source::Network(el) = source {
```

Comment:
So, nothing happens if the source is not `Network`, so there's no point in even calling the function if it's `Local`, right? Not a big deal but wonder if this function can not even care about the source and know that it's just `Network`?

Reply:
I'm not really sure what your asking but the reason the if let check is in here and not outside the call is otherwise you'd have to write it in every single function above and it adds a lot of noise. I think it's cleaner to just make it a noop when the source wasn't the network

Reply:
I'm going to start merging these but happy to open a new PR if you think it should be different